_posts/2021-03-17-encoding-effects-with-freer-simple.markdown

```diff
@@ -0,0 +1,474 @@
+---
+layout: post
+title: "Encoding Effects using freer-simple"
+author: Juan Pablo Royo Sales
+date: 2021-03-17
+excerpt: Organize your program encoding effects using freer-simple library. 
+categories: [haskell, fp, functional, effects, effects system, effects encoding, effect, encoding, type level programming, types, functional programming]
+tags: [haskell, fp, functional, effects, effects system, effects encoding, effect, encoding, type level programming, types, functional programming]
+permalink: /posts/2021-03-17-encoding-effects-freer-simple.html
```

Comment:
> The use case example is great since it uses a common use case in the software world. On the other hand, the test example is not great since it lacks the key part that is how do you use all these `mocked` interpreters. I would suggest adding an example which would be super useful and make the blog really great!
> 
> Good work!
> 
> Some general questions:
> 
> * In `runS3Storage`, why is the Handler needed ?

Because this is the implementation of the Algebra for uploading documents. In Testing section there is another Handler for the algebra that allows testing without connecting to AWS S3.

> * Effects with state how do you deal with them? For example, in the logging example, based on the previous log, add some metric like number of logs per second.

Good catch. In fact in testing section i am gathering all the logs and because of that i am using the `State [Text]`

> * Runtime sync/async exceptions (e.g. concurrent code). I saw that you catch the exception with a try but in the presence of async exception, how do you deal with them? I would be nice for a next blog to add an example on async computation with exceptions (for example, a timed coroutine).

Good catch. I was thinking on leaving that for a next blog. Async Exception, Concurrency, Retries and more advanced stuff.

> * Does the order of interpretation affect the semantics? In Algebraic Effect Handlers it does effect.

Yes. It affects.

> * In the examples, why do you need `State [Text]` in the `runTestcolog`?

It is because in testing you can test your logging lines gathering all of them in a list Text. So if you want to test Logging Algebra the way to do it is to gathering the lines and check against the final State of the list.